### PR TITLE
Use drawOnState() for buttons precessing and cache buttons' position for event handler loop

### PR DESCRIPTION
**Pull request by Districh-ru**
This PR affects most of the dialogs in the engine including the main menu.

It replaces `le.isMouseLeftButtonPressedInArea( button.area() ) ? button.drawOnPress() : button.drawOnRelease();` with `button.drawOnState( le.isMouseLeftButtonPressedInArea( buttonArea ) );` and do `buttonArea` cache not to call `.area()` with all its calculations on every loop.

It also modifies some dialogs that can be also called by the mouse right click to have its own event handler loop to reduce some checks (e.g. unit dialog, build dialog). And in some dialogs it skips the possibly unused buttons initialization making them `unique_ptr`, in example in Hero dialog: DISMISS and PATROL buttons.

While testing this PR you should check rendering of the buttons (press/release) and their interaction with the mouse cursor (right press, left click).

This PR does not fix rendering issues that are present in the master build.

### CODE REVIEW

**Review by ihhub — CHANGES_REQUESTED** [review]

File: src/fheroes2/battle/battle_dialogs.cpp [hunk 1]
```diff
@@ -336,15 +338,14 @@ namespace
             else if ( le.MouseClickLeft( optionAreas[6] ) ) {
                 return DialogAction::AudioSettings;
             }
-            if ( le.MouseClickLeft( optionAreas[7] ) ) {
+            else if ( le.MouseClickLeft( optionAreas[7] ) ) {
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/battle/battle_dialogs.cpp]: Hi, @ihhub, thanks for pointing this out!
I've fixed this and other places.

File: src/fheroes2/battle/battle_dialogs.cpp [hunk 2]
```diff
@@ -635,11 +638,13 @@ bool Battle::Arena::DialogBattleSummary( const Result & res, const std::vector<A
     int32_t buttonHorizontalMargin = 0;
     const int32_t buttonVerticalMargin = 5;
     std::unique_ptr<fheroes2::Button> buttonRestart;
+    std::unique_ptr<fheroes2::Rect> buttonRestartArea;
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/battle/battle_dialogs.cpp:641]: When the button is made its `area()` is often used.
Then should we just add `fheroes2::Rect _areaPressed;` and `fheroes2::Rect _areaReleased;` to the `ButtonBase` class?
Then we can make an inline method like:
```cpp
Rect & area()
{
    return _isPressed ? _areaPressed : _areaReleased;
}
```
And then remove all caches in of `area()` in the code.
What do you think, @ihhub?

[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/battle/battle_dialogs.cpp]: @ihhub, 
I've made the proposed change to cache area in `ButtonBase` class and the just return a const reference to it by `area()` call.

File: src/fheroes2/gui/ui_button.h [hunk 3]
```diff
@@ -164,19 +193,22 @@ namespace fheroes2
...
         void setICNIndexes( const uint32_t releasedIndex, const uint32_t pressedIndex )
         {
             _releasedIndex = releasedIndex;
             _pressedIndex = pressedIndex;
+
+            _updateButtonAreas();
         }
 
-    protected:
+    private:
```
Comment (ihhub) [inline src/fheroes2/gui/ui_button.h:208]: I suggest to keep the same section for these virtual functions for easier code reading. Having the section for base and child classes reduces confusion with understanding of sections.

The class is already marked as `final` so no other code can access protected section.

**Review by ihhub — COMMENTED** [review]
Hi @Districh-ru , I left just one small comment here. Would you mind please to take a look at it?

File: src/fheroes2/gui/ui_button.h (hunk 3 above)
Comment (Districh-ru) [reply src/fheroes2/gui/ui_button.h]: Sure!

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@Districh-ru , a bunch of thanks for these corrections!

**Pull request merged**